flocker/control/test/test_model_change.py

```diff
@@ -0,0 +1,104 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Whenever the configuration model changes we need to write code to
+upgrade the on-disk format from previous releases. This module will
+automatically detect such changes by failing a test, ensuring that upgrade
+code is always implemented when necessary.
+"""
+
+from json import loads, dumps
+from subprocess import check_output
+
+from pyrsistent import PRecord
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.python.filepath import FilePath
+from twisted.python.reflect import qual as fqpn
+
+from .._model import Deployment
+
+PERSISTED_MODEL = FilePath(__file__).sibling(b"persisted_model.json")
+
+# The class at the root of the configuration tree. This may need to be
+# changed if the configuration's root class changes.
+ROOT_CLASS = Deployment
+
+
+def generate_model():
+    """
+    Generate a data-structure that represents the current configuration
+    model.
+
+    Changes to output may require regenerating the persisted version
+    on-disk.
+    """
+    classes_result = {}
+    result = {u"root": fqpn(ROOT_CLASS),
+              u"classes": classes_result}
+    classes = {ROOT_CLASS}
+    while classes:
+        klass = classes.pop()
+        klass_name = fqpn(klass)
+        if klass_name in classes_result:
+            continue
+        record = None
+        if issubclass(klass, PRecord):
+            record = {u"category": u"record",
+                      u"fields": {}}
+            # XXX file issue with pyrsistent to add introspection API if I
+            # can't find anything in docs:
+            for name, field in klass._precord_fields.items():
+                record[u"fields"][name] = list(fqpn(cls) for cls in field.type)
+                for cls in field.type:
+                    classes.add(cls)
+        # XXX CheckedPMap and friends
```

Comment:
We currently need to support PRecord/PClass, CheckedPSet and CheckedPMap and that'll cover pretty much everything. If we change to a different class ... then that will be caught by the fact that the output changed. So we'd need to write more code, but we'd know that since test would fail.

Note that we don't cover "field types", there really isn't any such thing. A field points to types, it doesn't have a type.